chainer/training/extensions/evaluator.py

```diff
@@ -0,0 +1,168 @@
+    - Preparation routine can be customized, which is called before each
+      evaluation. It can be used, e.g., to initialize the state of stateful
+      recurrent networks.
+
+    There are two ways to modify the evaluation behavior besides setting a
+    custom evaluation function. One is by setting a custom evaluation loop via
+    the ``eval_loop`` argument. The other is by inheriting this class and
+    overriding the :meth:`evaluate` method. In latter case, users have to
+    create and handle a reporter object manually. Users also have to copy the
+    iterators before using them, in order to reuse them at the next time of
+    evaluation.
+
+    This extension is called at the end of each epoch by default.
+
+    Args:
+        iterator: Dataset iterator for the validation dataset. It can also be
+            a dictionary of iterators. If this is just an iterator, the
+            iterator is registered by the name ``'main'``.
+        target: Link object or a dictionary of links to evaluate. If this is
+            just a link object, the link is registered by the name ``'main'``.
+        eval_loop: Evaluation loop. This is a function that takes the evaluator
+            object as the argument, and returns a dictionary of statistics for
+            observed values. The default routine uses ``converter`` and
+            ``eval_func`` if specified.
+        converter: Converter function to build input arrays. If it is omitted,
+            :func:`~chainer.dataset.concat_examples` is used. If
+            ``eval_loop`` is specified, this argument is ignored and not used.
+        device: Device to which the training data is sent. Negative value
+            indicates the host memory (CPU). If ``eval_loop`` or ``converter``
+            is specified, this argument is ignored and not used.
+        eval_hook: Function to prepare for each evaluation process. It is
+            called at the beginning of the evaluation. The evaluator extension
+            object is passed at each call. If ``eval_loop`` is specified, this
+            argument is ignored and not used.
+        eval_func: Evaluation function called at each iteration. The target
+            link to evaluate as a callable is used by default. If ``eval_loop``
+            is specified, this argument is ignored and not used.
+
+    """
+    trigger = 1, 'epoch'
+    default_name = 'validation'
+    priority = extension.PRIORITY_WRITER
+
+    def __init__(self, iterator, target, eval_loop=None, converter=None,
+                 device=None, eval_hook=None, eval_func=None):
+        if isinstance(iterator, iterator_module.Iterator):
+            iterator = {'main': iterator}
+        self._iterators = iterator
+
+        if isinstance(target, link.Link):
+            target = {'main': target}
+        self._targets = target
+
+        self._eval_loop = eval_loop or _default_eval_loop(
+            self, converter, device, eval_hook, eval_func)
```

Comment:
Same discussion as [_update_func](https://github.com/pfnet/chainer/pull/1285/files#r69860219) is true of `_default_eval_loop`.